src/RulesProvider/AttributesRulesProvider.php

```diff
@@ -4,19 +4,18 @@
 
 namespace Yiisoft\Validator\RulesProvider;
 
+use Generator;
 use ReflectionAttribute;
 use ReflectionClass;
+use ReflectionException;
 use ReflectionObject;
 use ReflectionProperty;
 use Yiisoft\Validator\RuleInterface;
 use Yiisoft\Validator\RulesProviderInterface;
 
 final class AttributesRulesProvider implements RulesProviderInterface
 {
-    /**
-     * @var array<RuleInterface[]>|null
-     */
-    private ?array $rules = null;
+    private Generator|iterable|null $rules = null;
```

Comment:
```suggestion
    private Generator|iterable|null $rules;
```

Reply:
It's impossible to used typed property without default value out of the constructor.